### ISSUE TITLE
GGXF text format - grids in separate file

### ISSUE THREAD

**Issue by ccrook** [opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format#12]
This proposal is that in the GGXF text format, grid and polygon data be held in separate files.

The examples in the draft GGXF specification document use a text format for representing the GGXF format.  This format is valuable as it is immediately human readable, and can be used by producers to construct GGXF files. In particular the header content which requires much manual input by the producer.  It is now proposed that these be structured as YAML, a text format for data structures that has evolved from JSON as a machine readable format that is also easy for humans to read.  

However the GGXF text format examples also include grid data, and now also an optional extent polygon.  Neither of these will be created manually - they will be outputs from technical software and GIS systems.  Neither of them will (or practically can) be validated by a human reading them, except in very trivial cases. 

This potentially makes the format very unpalatable and very hard for a human reader to work with. For example a file containing a few small grid headers interspersed amongst millions  lines of grid data values.

I would like to propose that the grid and polygon data is not included in the YAML header descriptions, but instead held in individual data files which built alongside the YAML header.  The yaml header would just include references to these files (eg the filename of a grid file in the grid header section).  We could select a single supported format for each (grids and polygon definition) that is well supported by appropriate software.  For example the polygon could be WKT or WKB.  

There is a valid concern that if we split the GGXF text format into separate files then they may get separated.  One option to address this is to put all the files in a single zip file.  This is a commonly used pattern, for example in the Microsoft docx file.

In summary intention of this proposal is to create a human readable format for the bit that humans need to read and write, and not have to mix it up with the bit that software reads/writes and humans don't look at.

**Comment by RogerLott** [comment]
@ccrook is challenging a fundamental issue regarding the GGXF format - whether GGXF is a single file or a file collection. As I have stated in the PT meetings I have seen so many instances of attrocious data management with data separation and the necessary metadata being completely lost that I am not at all in favour of a multiple file capability - everything must be in a single file. 

Whether that file is in text or binary is in practice going to be determined by the size of the grids(s) and perhaps polygon(s) if this optional metadata is included. Of necessity most implementations will need to be binary. A text format is only going to be practical for very small grids, for showing to human readers small extracts of a large grid, or as an intermediate stage in file creation. There are precedents for file formats having text headers and binary data blocks (e.g. the SEG-Y format) but in my view this does not work especially well. I envisage the GGXF format having text and binary vaiants, with converters between the two.

The GGXF specification document is only intended for human readers, to define and illustrate the structure of a GGXF file. Of necessity that has to be wholly text. What goes into the GGXF specification document by way of illustrations (examples) should not drive the format itself. Using extracts of real-world examples with large grids in the draft specification document has perhaps unwittingly created this diversion into whether GGXF can be anything other than a single file. Drafting more appropriate informative examples that illuminate the file format is the way forward.

**Comment by desruisseaux** [comment]
I also believe that GGXF should be distributed as a single binary file. And I also agree that text files should be only an intermediate stage in the binary file creation. I see the text format like source code to be assembled in a single binary file by some "compiler". In this view, it would be okay if the text format is splitted in many files because those files would not be distributed; only the single binary file compiled from the text files would be distributed.

**Comment by kevinmkelly** [comment]
@RogerLott @desruisseaux @ccrook: I agree with this approach. In general, GGXF files should be single, binary files. Converters/compilers should be built to translate between binary and ascii representations. For very small GGXF files, these could be completely ascii and reader/writer routines built to work with them.

**Comment by ccrook** [comment]
It would certainly be possible to support both options to support a single text file as well.  Each grid could have either a gridData or gridSourceFile element depending on which is implemented.  For YAML that would require the correct syntax for gridData, depending on how we choose to organise it (eg whether the grid is flattened into a single linear array or structured as a multidimensional array).  

For non trivial nested grids it is unlikely that any human reader would get value from trying to view the text file.  I think it would only make sense as either multiple files or if viewed by some custom software for hiding the grid data (some code editors may handle this by folding the data - though the size of the file could be a bit daunting).

**Comment by rouault** [comment]
> It would certainly be possible to support both options to support a single text file as well.

Everything is doable, but I don't think it would be a good idea to have a text format as a legit alternative for the final exchange format. The less degrees of freedom offered, the better. If there's no appropriate standard binary carrier for everything that needs to be carried (metadata and grids), the idea of using a .zip file like @ccrook mentionned as being used for docx, ods, etc. (possibly not using the .zip extension itself, so folks aren't tempted to unzip it and distribute only parts of the content) could be reasonable. The drawback of using a .zip carrier is that for very large grids, seeking in a large file inside a .zip isn't efficient (you can't seek to an arbitrary location in a zip compressed file without uncompressing all the content of the file that is before the location you're interested in). Unless you use .zip without compression (the grid itself could use a standard compression mechanism of the grid format), but there might not be that many implementations that would know how to exploit this without actually unzipping the file.

Regarding polygon encoding, for a text format, GeoJSON would be a possibility (it could certainly be translated as YAML content, although that isn't very idiomatic)

**Comment by ccrook** [comment]
Almost everything is doable!  I agree that it doesn't make sense to have both text and ascii as the published and distributed model.

Maybe this does highlight a need to have clarity of what the purpose of the text format vs the binary format.  My preference is that this is a useful intermediate format to help publishers create and maintain GGXF data sets.   This can be supported by  a standard validator/translator to build the binary format.  It provides a straightforward consistent gateway from the often custom scientific software used to analyse and build deformation grids to the GGXF format.

**Comment by ccrook** [comment]
This is not currently supported in the GGXF specification, though it does now support a gridData element which separates the data from the YAML headers, which addresses on of the concerns of this issue.  

It is implemented in the POC YAML implementation using a dataSource attribute in grid.  In the POC this uses the GDAL data source format to specify the grid source, and it may need further enhancement to support eg selecting specific parameters from the source grid

Ideally if supported this would be a normative specification to ensure consistency.  

This could be done by defining a specific format, eg simple CSV file lat,lon,param1,param2,...

Alternatively the specification could support an implementation specific dataSource parameter and leave it up the YAML implementation to choose how to support it.

Also it could support both options - ie a YAML implementation must support CSV or selected format, and support alternative formats using a defined extension syntax (eg "gdal:_gdal_data_source_parameters_").

To be resolved.

**Comment by ccrook** [comment]
The use of a dataSource attribute in YAML to identify an application specific encoding of external data is included in the specification.  The encoding of an external grid source is not defined in the application.  

As an example the [proof of concept code](https://github.com/opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format/tree/531a1f5cbfbe271da24191e9eb262a06e9b35138) implements this using a structured source type as

```
dataSource:
    sourceType: GDAL
    gdalSource: "GTIFF_DIR:2:nz_linz_nzgd2000-ndm-grid02.tif"

dataSource:
    sourceType: CSV
    csvFilename: data.csv
    interpolationCoordFields: [lat, lon]
    parameterFields: [de, dn, du]
```

**Comment by kevinmkelly** [comment]
Does the first dataSource require a URL or file location of some sort?

**Comment by ccrook** [comment]
@kevinmkelly That example is a GDAL connection string.  The actual content depends on what GDAL driver is being invoked, in this case GTIFF_DIR which specifies that it is a GeoTIFF file (and 2 specifying the grid within it).   However a driver using an online source would have a url.

As far as the GGXF specification is concerned the content of dataSource is not specified, another software could use a different data source.

The main motivation for using dataSource is that very large grids are challenging for YAML readers, so for practical reasons something that is more suitalbe for loading grids is a good idea.  GGXF could specify one required external dataSource type (eg CSV which is fairly generic).  But definitely we want to avoid supporting the plethora of grid types that there already are - that would be completely counter productive!!   

As far as the GDAL dataSource is concerned this definitely should not be a GGXF specification - it is a "feature/distraction" of the POC code that I put in mainly to simplify importing existing data sets without having to do a lot of translating of grids.

For the current version of the specification we are simply saying that it an attribute that individual applications may choose to use as a customisation of the YAML format.

**Comment by desruisseaux** [comment]
Should we include a non-normative part giving examples of what may look like with some formats? Given that GeoTIFF and netCDF-3 (a.k.a. "classic") are two OGC standards, we could provide two examples using them. Maybe something like:

```
dataSource:
    format: GeoTIFF
    path: nz_linz_nzgd2000-ndm-grid02.tif
    imageIndex: 2

dataSource:
    format: netCDF
    path: nz_linz_nzgd2000-ndm-grid02.nc
    variable: foo
```

**Comment by ccrook** [comment]
Difficult to know how far to go with examples.  If we are putting much detail in then, unless we are specifying a specific format, we need to show that different software may use different options.  If we were specifying then I would choose to just include a text based format to complement YAML, rather than one, or many, of the formats that we are hoping to replace for geodetic data.  

If we did have a non-normative section we could show your GeoTIFF example alongside my GDAL example to emphasize that different software may define this differently?   

There is already an example of using dataSource in E5.   (Based on a previous implementation in the POC)

The introductory text of this example should perhaps restate that the content of the dataSource attribute is application specific and show how it might be used in a different software with something like your suggestion.  That way a user won't copy this example and then not understand why it doesn't work with their GGXF software.

**Comment by RogerLott** [comment]
A fundamental tenet of GGXF is that header and grid data are in one file. dataSource is in the specification to assist producers in achieving this. Whilst dataSource as a keyword is part of the GGXF specification, the attributes of dataSource and their application are not part of of the GGXF specification. As @ccrook suggests, we can restate this in E.5, but why should we detail anything that is not in the spec?

**Comment by RogerLott** [comment]
ggxf-csv added to v1.0, available at https://github.com/opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format/blob/master/specification/GGXF%20v1-0%20OGC-22-051_full.pdf
